File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/ui/client/pages/CouponManagementPage.java
package com.mycompany.myshop.testkit.driver.adapter.ui.client.pages;

import com.mycompany.myshop.testkit.driver.adapter.shared.client.playwright.PageClient;
import com.mycompany.myshop.testkit.common.Converter;
import com.mycompany.myshop.testkit.driver.port.dtos.BrowseCouponsResponse;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

public class CouponManagementPage extends BasePage {
    private static final String COUPON_CODE_INPUT_SELECTOR = "[aria-label=\"Coupon Code\"]";
    private static final String DISCOUNT_RATE_INPUT_SELECTOR = "[aria-label=\"Discount Rate\"]";
    private static final String VALID_FROM_INPUT_SELECTOR = "[aria-label=\"Valid From\"]";
    private static final String VALID_TO_INPUT_SELECTOR = "[aria-label=\"Valid To\"]";
    private static final String USAGE_LIMIT_INPUT_SELECTOR = "[aria-label=\"Usage Limit\"]";
    private static final String PUBLISH_COUPON_BUTTON_SELECTOR = "[aria-label=\"Create Coupon\"]";

    // Selectors for browsing coupons
    private static final String COUPONS_TABLE_SELECTOR = "[aria-label=\"Coupons Table\"]";

    // Time format constants
    private static final String TIME_MIDNIGHT = "T00:00";
    private static final String TIME_END_OF_DAY = "T23:59";

    // Table selectors
    private static final String TABLE_CELL_CODE_SELECTOR = "table.table tbody tr td:nth-child(1)";
    private static final String TABLE_CELL_DISCOUNT_SELECTOR = "table.table tbody tr td:nth-child(2)";
    private static final String TABLE_CELL_VALID_FROM_SELECTOR = "table.table tbody tr td:nth-child(3)";
    private static final String TABLE_CELL_VALID_TO_SELECTOR = "table.table tbody tr td:nth-child(4)";
    private static final String TABLE_CELL_USAGE_LIMIT_SELECTOR = "table.table tbody tr td:nth-child(5)";
    private static final String TABLE_CELL_USED_COUNT_SELECTOR = "table.table tbody tr td:nth-child(6)";

    // Display text constants
    private static final String PERCENT_SYMBOL = "%";
    private static final String TEXT_IMMEDIATE = "Immediate";
    private static final String TEXT_NEVER = "Never";
    private static final String TEXT_UNLIMITED = "Unlimited";

    public CouponManagementPage(PageClient pageClient) {
        super(pageClient);
    }

    public void inputCouponCode(String couponCode) {
        pageClient.fill(COUPON_CODE_INPUT_SELECTOR, couponCode);
    }

    public void inputDiscountRate(String discountRate) {
        pageClient.fill(DISCOUNT_RATE_INPUT_SELECTOR, discountRate);
    }

    public void inputValidFrom(String validFrom) {
        var datetimeValue = getValidFromDateTimeString(validFrom);
        pageClient.fill(VALID_FROM_INPUT_SELECTOR, datetimeValue);
    }

    public void inputValidTo(String validTo) {
        var datetimeValue = getValidToDateTimeString(validTo);
        pageClient.fill(VALID_TO_INPUT_SELECTOR, datetimeValue);
    }

    public void inputUsageLimit(String usageLimit) {
        pageClient.fill(USAGE_LIMIT_INPUT_SELECTOR, usageLimit);
    }

    public void clickPublishCoupon() {
        pageClient.click(PUBLISH_COUPON_BUTTON_SELECTOR);
    }

    public List<BrowseCouponsResponse.CouponDto> readCoupons() {
        // Wait for table to appear/refresh after operations like PublishCoupon
        pageClient.waitForVisible(COUPONS_TABLE_SELECTOR);

        var coupons = new ArrayList<BrowseCouponsResponse.CouponDto>();
        // Use readAllTextContentsWithoutWait to avoid strict mode violations
        // These selectors intentionally match multiple elements (one per table row)
        var codes = pageClient.readAllTextContentsNoWait(TABLE_CELL_CODE_SELECTOR);

        // If no codes found, table is empty
        if (codes.isEmpty()) {
            return new ArrayList<>();
        }

        var discountRates = pageClient.readAllTextContents(TABLE_CELL_DISCOUNT_SELECTOR);
        var validFroms = pageClient.readAllTextContents(TABLE_CELL_VALID_FROM_SELECTOR);
        var validTos = pageClient.readAllTextContents(TABLE_CELL_VALID_TO_SELECTOR);
        var usageLimits = pageClient.readAllTextContents(TABLE_CELL_USAGE_LIMIT_SELECTOR);
        var usedCounts = pageClient.readAllTextContents(TABLE_CELL_USED_COUNT_SELECTOR);

        var rowCount = codes.size();

        // Double-check we have data before trying to access it
        // Also verify all columns have the same row count (handles empty tables or malformed data)
        if (rowCount == 0 || discountRates.size() != rowCount || validFroms.size() != rowCount
                || validTos.size() != rowCount || usageLimits.size() != rowCount || usedCounts.size() != rowCount) {
            return new ArrayList<>();
        }

        // Build coupon objects from the collected data
        for (int i = 0; i < rowCount; i++) {
            var code = codes.get(i).trim();
            var discountRateText = discountRates.get(i).trim().replace(PERCENT_SYMBOL, "");
            var validFromText = validFroms.get(i).trim();
            var validToText = validTos.get(i).trim();
            var usageLimitText = usageLimits.get(i).trim();
            var usedCountText = usedCounts.get(i).trim();

            var coupon = BrowseCouponsResponse.CouponDto.builder()
                    .code(code)
                    .discountRate(parseDiscountRate(discountRateText))
                    .validFrom(toInstant(validFromText))
                    .validTo(toInstant(validToText))
                    .usageLimit(parseUsageLimit(usageLimitText))
                    .usedCount(Converter.toInteger(usedCountText))
                    .build();

            coupons.add(coupon);
        }

        return coupons;
    }

    private static String getValidFromDateTimeString(String validFrom) {
        if(validFrom == null || validFrom.isEmpty()) {
            return "";
        }

        String dateOnly = validFrom.substring(0, 10);
        return dateOnly + TIME_MIDNIGHT;
    }

    private static String getValidToDateTimeString(String validTo) {
        if(validTo == null || validTo.isEmpty()) {
            return "";
        }

        String dateOnly = validTo.substring(0, 10);
        return dateOnly + TIME_END_OF_DAY;
    }

    private double parseDiscountRate(String text) {
        if(text == null || text.isEmpty()) {
            return 0.00;
        }

        var value = Converter.toDouble(text);
        return value == null ? 0.00 : value / 100.0;
    }

    private Instant toInstant(String text) {
        return Converter.toInstant(text, TEXT_IMMEDIATE, TEXT_NEVER);
    }

    private Integer parseUsageLimit(String text) {
        return Converter.toInteger(text, TEXT_UNLIMITED);
    }
}

File: system-test/dotnet/Driver.Adapter/Ui/Client/Pages/CouponManagementPage.cs
using Driver.Adapter.Shared.Client.Playwright;
using Common;
using Driver.Port.Dtos;

namespace Driver.Adapter.Ui.Client.Pages;

public class CouponManagementPage : BasePage
{
    private const string CouponCodeInputSelector = "[aria-label=\"Coupon Code\"]";
    private const string DiscountRateInputSelector = "[aria-label=\"Discount Rate\"]";
    private const string ValidFromInputSelector = "[aria-label=\"Valid From\"]";
    private const string ValidToInputSelector = "[aria-label=\"Valid To\"]";
    private const string UsageLimitInputSelector = "[aria-label=\"Usage Limit\"]";
    private const string PublishCouponButtonSelector = "[aria-label=\"Create Coupon\"]";

    // Selectors for browsing coupons
    private const string CouponsTableSelector = "[aria-label=\"Coupons Table\"]";

    // Time format constants
    private const string TimeMidnight = "T00:00";
    private const string TimeEndOfDay = "T23:59";

    // Table selectors
    private const string TableCellCodeSelector = "table.table tbody tr td:nth-child(1)";
    private const string TableCellDiscountSelector = "table.table tbody tr td:nth-child(2)";
    private const string TableCellValidFromSelector = "table.table tbody tr td:nth-child(3)";
    private const string TableCellValidToSelector = "table.table tbody tr td:nth-child(4)";
    private const string TableCellUsageLimitSelector = "table.table tbody tr td:nth-child(5)";
    private const string TableCellUsedCountSelector = "table.table tbody tr td:nth-child(6)";

    // Display text constants
    private const string PercentSymbol = "%";
    private const string TextImmediate = "Immediate";
    private const string TextNever = "Never";
    private const string TextUnlimited = "Unlimited";

    public CouponManagementPage(PageClient pageClient) : base(pageClient)
    {
    }

    public async Task InputCouponCodeAsync(string? couponCode)
    {
        await PageClient.FillAsync(CouponCodeInputSelector, couponCode);
    }

    public async Task InputDiscountRateAsync(string? discountRate)
    {
        await PageClient.FillAsync(DiscountRateInputSelector, discountRate);
    }

    public async Task InputValidFromAsync(string? validFrom)
    {
        var datetimeValue = GetValidFromDateTimeString(validFrom);
        await PageClient.FillAsync(ValidFromInputSelector, datetimeValue);
    }

    public async Task InputValidToAsync(string? validTo)
    {
        var datetimeValue = GetValidToDateTimeString(validTo);
        await PageClient.FillAsync(ValidToInputSelector, datetimeValue);
    }

    public async Task InputUsageLimitAsync(string? usageLimit)
    {
        await PageClient.FillAsync(UsageLimitInputSelector, usageLimit);
    }

    public async Task ClickPublishCouponAsync()
    {
        await PageClient.ClickAsync(PublishCouponButtonSelector);
    }

    public async Task<List<CouponDto>> ReadCouponsAsync()
    {
        // Wait for table to appear/refresh after operations like PublishCoupon
        await PageClient.WaitForVisibleAsync(CouponsTableSelector);

        var coupons = new List<CouponDto>();
        // Use readAllTextContents to avoid strict mode violations
        // These selectors intentionally match multiple elements (one per table row)
        var codes = await PageClient.ReadAllTextContentsAsync(TableCellCodeSelector);

        // If no codes found, table is empty
        if (codes.Count == 0)
        {
            return new List<CouponDto>();
        }

        var discountRates = await PageClient.ReadAllTextContentsAsync(TableCellDiscountSelector);
        var validFroms = await PageClient.ReadAllTextContentsAsync(TableCellValidFromSelector);
        var validTos = await PageClient.ReadAllTextContentsAsync(TableCellValidToSelector);
        var usageLimits = await PageClient.ReadAllTextContentsAsync(TableCellUsageLimitSelector);
        var usedCounts = await PageClient.ReadAllTextContentsAsync(TableCellUsedCountSelector);

        var rowCount = codes.Count;

        // Double-check we have data before trying to access it
        // Also verify all columns have the same row count (handles empty tables or malformed data)
        if (rowCount == 0 || discountRates.Count != rowCount || validFroms.Count != rowCount
                || validTos.Count != rowCount || usageLimits.Count != rowCount || usedCounts.Count != rowCount)
        {
            return new List<CouponDto>();
        }

        // Build coupon objects from the collected data
        for (int i = 0; i < rowCount; i++)
        {
            var code = codes[i].Trim();
            var discountRateText = discountRates[i].Trim().Replace(PercentSymbol, "");
            var validFromText = validFroms[i].Trim();
            var validToText = validTos[i].Trim();
            var usageLimitText = usageLimits[i].Trim();
            var usedCountText = usedCounts[i].Trim();

            var coupon = new CouponDto
            {
                Code = code,
                DiscountRate = ParseDiscountRate(discountRateText),
                ValidFrom = ToDateTime(validFromText),
                ValidTo = ToDateTime(validToText),
                UsageLimit = ParseUsageLimit(usageLimitText),
                UsedCount = Converter.ToInteger(usedCountText) ?? 0
            };

            coupons.Add(coupon);
        }

        return coupons;
    }

    private static string GetValidFromDateTimeString(string? validFrom)
    {
        if (string.IsNullOrEmpty(validFrom))
        {
            return "";
        }

        // Extract date from ISO 8601 format (2024-08-31T00:00:00Z -> 2024-08-31)
        // Then convert to datetime-local format (2024-08-31T00:00) for HTML input
        string dateOnly = validFrom.Substring(0, 10); // YYYY-MM-DD
        return dateOnly + TimeMidnight;
    }

    private static string GetValidToDateTimeString(string? validTo)
    {
        if (string.IsNullOrEmpty(validTo))
        {
            return "";
        }

        // Extract date from ISO 8601 format (2024-08-31T23:59:59Z -> 2024-08-31)
        // Then convert to datetime-local format (2024-08-31T23:59) for HTML input
        string dateOnly = validTo.Substring(0, 10); // YYYY-MM-DD
        return dateOnly + TimeEndOfDay;
    }

    private static decimal ParseDiscountRate(string? text)
    {
        var value = Converter.ToDecimal(text);
        return value == null ? 0.0m : value.Value / 100.0m; // Convert percentage to decimal
    }

    private static DateTime? ToDateTime(string text)
    {
        return Converter.ToDateTime(text, TextImmediate, TextNever);
    }

    private static int? ParseUsageLimit(string? text)
    {
        return Converter.ToInteger(text, TextUnlimited);
    }
}
